### PR TITLE
patch: [TECH-7725] Change COMMAND_OUTPUT to fallback to posix-standard in worst case

### PR DESCRIPTION
**Pull request by joseph-galindo**
[QA: None]

Context in https://compass-tech.atlassian.net/browse/TECH-7725

Particularly this fixes things like:
```
npm ERR! path /home/myuser/myproject/node_modules/aws-sdk
npm ERR! command failed
npm ERR! command sh -c node scripts/warn-maintenance-mode.js
npm ERR! /home/myuser/.basher/cellar/packages/UrbanCompass/hnvm/lib/hnvm/config.sh: line 215: /dev/stdout: No such device or address
```

This will now try to check stdout target existence, before using it as a redirect target

### CODE REVIEW

File: lib/hnvm/config.sh [hunk 1]
```diff
@@ -12,7 +12,34 @@ while [ -h "$source" ]; do # resolve $source until the file is no longer a symli
...
+if [[ -e "$HNVM_OUTPUT_DESTINATION" ]]; then
+  COMMAND_OUTPUT="$HNVM_OUTPUT_DESTINATION"
+fi
+
+# If COMMAND_OUTPUT is STILL empty, fall back to posix-standard /dev/null
+if [[ -z "$COMMAND_OUTPUT" ]]; then
+  echo "WARNING: Could not find a valid stdout redirect target!"
+  echo "WARNING: Further HNVM output will be redirected to '/dev/null'"
+  COMMAND_OUTPUT="/dev/null"
+fi
```
Comment (josephschmitt) [inline lib/hnvm/config.sh:40]: Why wouldn't these be `elif` and `else`?

File: lib/hnvm/config.sh [hunk 2]
```diff
@@ -12,7 +12,34 @@ while [ -h "$source" ]; do # resolve $source until the file is no longer a symli
...
+  COMMAND_OUTPUT="/dev/fd/1"
+fi
+
+if [[ -e "/dev/stdout" ]]; then
+  COMMAND_OUTPUT="/dev/stdout"
+fi
+
+if [[ -e "$HNVM_OUTPUT_DESTINATION" ]]; then
+  COMMAND_OUTPUT="$HNVM_OUTPUT_DESTINATION"
+fi
```
Comment (joseph-galindo) [inline lib/hnvm/config.sh:33]: This is typically used for `echo "foo" >> $HNVM_OUTPUT_DESTINATION`, where `>>` will create the file if it doesn't exist.

So what may be more fitting is
```bash
# if HNVM_OUTPUT_DESTINATION is a nonzero len string, use it, we will handle creating the file (if needed) then appending
if [[ -n "$HNVM_OUTPUT_DESTINATION" ]]; then
  COMMAND_OUTPUT="$HNVM_OUTPUT_DESTINATION"
fi
```

File: lib/hnvm/config.sh (hunk 1 above)
Comment (joseph-galindo) [reply lib/hnvm/config.sh:40]: No strong reason, I'm just used to structuring this type of hierarchy stuff in this way. I can change it over to elif, else

File: madeupthing [hunk 3]
```diff
@@ -0,0 +1 @@
+foo
```
Comment (joseph-galindo) [inline madeupthing:1]: removing, I was testing redirects

File: lib/hnvm/config.sh (hunk 1 above)
Comment (joseph-galindo) [reply lib/hnvm/config.sh]: done

File: lib/hnvm/config.sh [hunk 4]
```diff
@@ -12,7 +12,42 @@ while [ -h "$source" ]; do # resolve $source until the file is no longer a symli
...
-export COMMAND_OUTPUT=${HNVM_OUTPUT_DESTINATION:-/dev/stdout}
+COMMAND_OUTPUT=""
+
+# Try aggressively to find an environment-supported, redirectable, stdout fd descriptor
+#
+# In nodejs spawned subprocess (think npm pre/postinstall scripts), stdio can be
+# provided as sockets, which is not able to do simple '>>' bash redirection
+# https://nodejs.org/api/child_process.html#optionsstdio
+# "These are not actual Unix pipes and therefore the child process can not use them by their
+# descriptor files, e.g. /dev/fd/2 or /dev/stdout."
```
Comment (joseph-galindo) [inline lib/hnvm/config.sh:23]: Relevant art
- https://stackoverflow.com/questions/72858721/error-no-such-device-or-address-from-child-process-exec-pipe-helps-why
- https://stackoverflow.com/questions/37618443/why-cant-bash-recognize-the-existence-of-a-socket-file
- Especially this explanation: https://stackoverflow.com/a/72906798
- nodejs docs https://nodejs.org/api/child_process.html#optionsstdio

I reworked things to try to entirely avoid using sockets as targets of our bash redirects. I also added tests to make sure hnvm still works (by sending to `/dev/null`), in the rare case that the only other options are sockets.

---
Originally noticed in https://compass-tech.atlassian.net/browse/TECH-7725

I noticed similar crash behavior to the ticket, [in CI test runs for this pr](https://app.circleci.com/pipelines/github/UrbanCompass/hnvm/104/workflows/830b67dd-11bb-46ab-82c3-023130e636b5/jobs/149?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary), so I added the new test to make sure the problem is resolved in our CI runs

File: test/utils.js [hunk 5]
```diff
@@ -30,6 +64,8 @@ function createTestContext() {
       fs.writeFileSync(path.join(cwdDir, 'package.json'), JSON.stringify(json))
     },
     cleanup() {
+      // close() will close the server bound to the socket file, and delete the socket file
```
Comment (joseph-galindo) [inline test/utils.js:67]:
```suggestion
      // close() will close the server bound to the socket file, and terminate any recorded socket connections
```

File: lib/hnvm/config.sh [hunk 6]
```diff
@@ -12,7 +12,42 @@ while [ -h "$source" ]; do # resolve $source until the file is no longer a symli
...
+#
+# For command output shell redirection, hierarchy is (most to least priority)
+# - $HNVM_OUTPUT_DESTINATION - arbitrary path given as env var
+# - /dev/stdout - convenience symlink to fd 1 - not fully posix portable
+# - /dev/fd/1 - on macos, the character special device (c in `ls`) that is the stdout stream
+# - /dev/null - only used as fallback, if none of the above exist
```
Comment (josephschmitt) [inline lib/hnvm/config.sh:29]: I think all of the comments here are actually making the code harder to read. I think maybe we pull them out and just add this detail to the PR as comments. You can always use git blame to see the discussion and reasoning. At most I'd be comfortable with Pointing to a ticket, PR, or stackoverflow file or something explaining the issue

Comment (joseph-galindo) [reply lib/hnvm/config.sh:29]: Sounds good, I will rework this

File: lib/hnvm/config.sh [hunk 7]
```diff
@@ -12,7 +12,42 @@ while [ -h "$source" ]; do # resolve $source until the file is no longer a symli
...
+if [[ -n "$HNVM_OUTPUT_DESTINATION" && ! -S "$HNVM_OUTPUT_DESTINATION" ]]; then
```
Comment (joseph-galindo) [inline lib/hnvm/config.sh:30]: Try aggressively to find an environment-supported, redirectable, stdout fd descriptor

In nodejs spawned subprocess (think npm pre/postinstall scripts), stdio can be
provided as sockets, which is not able to do simple '>>' bash redirection
https://nodejs.org/api/child_process.html#optionsstdio
> "These are not actual Unix pipes and therefore the child process can not use them by their descriptor files, e.g. /dev/fd/2 or /dev/stdout."

For command output shell redirection, hierarchy is (most to least priority)
- $HNVM_OUTPUT_DESTINATION - arbitrary path given as env var
- /dev/stdout - convenience symlink to fd 1 - not fully posix portable
- /dev/fd/1 - on macos, the character special device (c in `ls`) that is the stdout stream
- /dev/null - only used as fallback, if none of the above exist

File: lib/hnvm/config.sh [hunk 8]
```diff
@@ -12,7 +12,42 @@ while [ -h "$source" ]; do # resolve $source until the file is no longer a symli
...
+  # If this is a non-zero string, and does not point to an existing socket, we will use it as-is.
+  # Unlike the fallbacks, don't check for existence - we support using a path that does not exist
+  # Reason is, '>>' redirect will create the file if it doesn't already exist
+  COMMAND_OUTPUT="$HNVM_OUTPUT_DESTINATION"
```
Comment (joseph-galindo) [inline lib/hnvm/config.sh:34]: If this is a non-zero string, and does not point to an existing socket, we will use it as-is.
  Unlike the fallbacks, don't check for existence - we support using a path that does not exist
  Reason is, '>>' redirect will create the file if it doesn't already exist, so the file at `$HNVM_OUTPUT_DESTINATION` does **not** have to exist prior to running

File: lib/hnvm/config.sh [hunk 9]
```diff
@@ -12,7 +12,42 @@ while [ -h "$source" ]; do # resolve $source until the file is no longer a symli
...
+elif [[ -e "/dev/stdout" && -w "/dev/stdout" && ! -S "/dev/stdout" ]]; then
+  # Only use this when the fd target exists, is writable, and is NOT a socket
+  # 'echo "foo" >> $target' redirect will not work, if $target is a unix IPC socket
+  COMMAND_OUTPUT="/dev/stdout"
```
Comment (joseph-galindo) [inline lib/hnvm/config.sh:38]: Only use this when the fd target exists, is writable, and is NOT a socket
'echo "foo" >> $target' redirect will not work, if $target is a unix IPC socket

File: lib/hnvm/config.sh [hunk 10]
```diff
@@ -12,7 +12,42 @@ while [ -h "$source" ]; do # resolve $source until the file is no longer a symli
...
+elif [[ -e "/dev/fd/1" && -w "/dev/fd/1" && ! -S "/dev/stdout" ]]; then
+  # Only use this when the fd target exists, is writable, and is NOT a socket
+  # 'echo "foo" >> $target' redirect will not work, if $target is a unix IPC socket
+  COMMAND_OUTPUT="/dev/fd/1"
```
Comment (joseph-galindo) [inline lib/hnvm/config.sh:42]: Only use this when the fd target exists, is writable, and is NOT a socket
'echo "foo" >> $target' redirect will not work, if $target is a unix IPC socket

File: lib/hnvm/config.sh (hunk 6 above)
Comment (joseph-galindo) [reply lib/hnvm/config.sh]: done

File: lib/hnvm/config.sh (hunk 7 above)
Comment (joseph-galindo) [reply lib/hnvm/config.sh:19]: The main motive for this hierarchy, is to prevent this from ever happening:

- https://compass-tech.atlassian.net/browse/TECH-7725
- https://app.circleci.com/pipelines/github/UrbanCompass/urbancompass/993565/workflows/454d4c7d-e413-47da-a495-1cfd7d0836cf
```
npm ERR! path /home/myuser/myproject/node_modules/aws-sdk
npm ERR! command failed
npm ERR! command sh -c node scripts/warn-maintenance-mode.js
npm ERR! /home/myuser/.basher/cellar/packages/UrbanCompass/hnvm/lib/hnvm/config.sh: line 215: /dev/stdout: No such device or address
```

File: lib/hnvm/config.sh [hunk 11]
```diff
@@ -12,7 +12,24 @@ while [ -h "$source" ]; do # resolve $source until the file is no longer a symli
 done
 script_dir="$( cd -P "$( dirname "$source" )" >/dev/null 2>&1 && pwd )"
 
-export COMMAND_OUTPUT=${HNVM_OUTPUT_DESTINATION:-/dev/stdout}
+COMMAND_OUTPUT=""
+
+# Context for why we try multiple output redirect targets:
```
Comment (josephschmitt) [inline lib/hnvm/config.sh:17]: 👍

File: test/node-subprocess.test.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  })
+
+  it('should fallback to using "/dev/null" if HNVM_OUTPUT_DESTINATION is a socket', () => {
+    const result = context.execFileSyncWithSocketOutput(
+      context.binaries.node,
+      ['-p', '"Hello, World!"']
+    )
+
+    expect(result).toContain('Hello, World!')
+    expect(result).toContain(
```
Comment (josephschmitt) [inline test/node-subprocess.test.js:25]: Nice

**Review by josephschmitt — APPROVED** [review]
